### PR TITLE
Implement Contact wrench Cone in math component 

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
The math is taken from [Stability of surface contacts for humanoid robots: Closed-form formulae of the Contact Wrench Cone for rectangular support areas](https://ieeexplore.ieee.org/document/7139910)

The PR is required to implement the Feasibility wrench constraint in TSID

cc @Giulero

### CODE REVIEW

File: src/Math/include/BipedalLocomotion/Math/ContactWrenchCone.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ * require CoP inside the support area and a bounded yaw torque. The ContactWrenchCone aims to
+ * compute the polyhedral approximation of \f$ | f ^t | \le \mu f^c \cdot n \f$ by spitting the base
+ * of the cone into slices while considering the contact torque constraints. The class implements
+ * the equations presented in [_Stability of Surface Contacts for Humanoid Robots: Closed-Form
+ * Formulae of the Contact Wrench Cone for Rectangular Support Areas_
+ * paper](https://ieeexplore.ieee.org/document/7139910). However differently from the original
+ * work, the origin of the frame attached to the contact surface is not required to be in the
+ * center of the surface.
+ * @note The ContactWrenchCone class express the constraints in body representation (so called right
+ * trivialization). Please check
```
Comment (S-Dafarra) [inline src/Math/include/BipedalLocomotion/Math/ContactWrenchCone.h:42]: Isn't body the left trivialization?

File: src/Math/include/BipedalLocomotion/Math/Skew.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ */
+template <typename _Scalar>
+typename std::enable_if<std::is_arithmetic<_Scalar>::value, Eigen::Matrix<_Scalar, 2, 2>>::type
+skew(const _Scalar v)
+{
+    return (Eigen::Matrix<_Scalar, 2, 2>() << _Scalar(0.), -v, v, _Scalar(0.)).finished();
+}
+
+/**
+ * Return a 3x3 skew matrix given a scalar.
```
Comment (S-Dafarra) [inline src/Math/include/BipedalLocomotion/Math/Skew.h:40]: It is not a scalar.

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,8 @@ All notable changes to this project are documented in this file.
 ### Added
 - Implement `CubicSpline` class (https://github.com/dic-iit/bipedal-locomotion-framework/pull/344)
 - Implement `PWM` control in RobotControl class (https://github.com/dic-iit/bipedal-locomotion-framework/pull/346)
+- Implement `ContactWrenchCone` class in Math component (https://github.com/dic-iit/bipedal-locomotion-framework/pull/352)
+- Implement `skew` function in Math component (https://github.com/dic-iit/bipedal-locomotion-framework/pull/352)
```
Comment (prashanthr05) [inline CHANGELOG.md:9]: Just FYI, `skew` exists also in manif for both 2d and 3d cases, 
https://github.com/artivis/manif/blob/7c6b8b2dab6d940204136a5b280db1c243d72c4a/include/manif/impl/eigen.h#L131-L166

File: src/Math/tests/FrictionConesTest.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
+/**
+ * @file LinearizedFrictionConeTest.cpp
```
Comment (prashanthr05) [inline src/Math/tests/FrictionConesTest.cpp:2]:
```suggestion
 * @file FrictionConesTest.cpp
```

**Review by S-Dafarra — COMMENTED** [review]
I would add some test cases where you test the constraint with values you are sure are inside, and values for which you are sure are outside. I have seen you have tested with respect to what you have computed in Matlab, but you could have done the same mistake twice. Also, if you need to edit the constraint by adding or removing lines, that part would break.

File: src/Math/include/BipedalLocomotion/Math/ContactWrenchCone.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+ * A surface remains fixed with respect the environment if the contact wrench lies in a cone
+ * described by
+ * \f[
+ * f ^c \cdot n > 0 \quad | f ^t | \le \mu f^c \cdot n
+ * \f]
+ *  where \f$ f^c \f$ is the
+ * contact force, \f$ n \f$ is the vector normal to the contact surface. \f$ f^t \f$ is the
+ * tangential force to the contact surface and \f$ \mu \f$ is the friction parameter. In addition we
+ * require CoP inside the support area and a bounded yaw torque. The ContactWrenchCone aims to
+ * compute the polyhedral approximation of \f$ | f ^t | \le \mu f^c \cdot n \f$ by spitting the base
```
Comment (prashanthr05) [inline src/Math/include/BipedalLocomotion/Math/ContactWrenchCone.h:34]:
```suggestion
 * compute the polyhedral approximation of \f$ | f ^t | \le \mu f^c \cdot n \f$ by splitting the base
```

File: src/Math/include/BipedalLocomotion/Math/ContactWrenchCone.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+
+namespace BipedalLocomotion
+{
+namespace Math
+{
+
+/**
+ * ContactWrenchCone computes the polyhedral approximation of the contact wrench friction cone.
+ * A surface remains fixed with respect the environment if the contact wrench lies in a cone
```
Comment (prashanthr05) [inline src/Math/include/BipedalLocomotion/Math/ContactWrenchCone.h:25]:
```suggestion
 * A surface remains fixed with respect to the environment if the contact wrench lies in a cone
```

File: src/Math/include/BipedalLocomotion/Math/ContactWrenchCone.h (hunk 1 above)
Comment (prashanthr05) [inline src/Math/include/BipedalLocomotion/Math/ContactWrenchCone.h:42]:
```suggestion
 * @note The ContactWrenchCone class express the constraints in body representation (so called left
 * trivialization). Please check
```

Body = left
Spatial = right

File: src/Math/src/ContactWrenchCone.cpp [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    AYawTorque.row(6) << -halfRectangleWidth, halfRectangleLength,
+        -(halfRectangleWidth + halfRectangleLength) * staticFrictionCoefficient,
+        -staticFrictionCoefficient, staticFrictionCoefficient, 1;
+
+    AYawTorque.row(7) << -halfRectangleWidth, -halfRectangleLength,
+        -(halfRectangleWidth + halfRectangleLength) * staticFrictionCoefficient,
+        -staticFrictionCoefficient, -staticFrictionCoefficient, 1;
+
+    // AYawTorque expects a wrench expressed in the center of the contact surface. The following
+    // adjoin transform brings the wrench expressed in the frame attached to the contact surface in
```
Comment (prashanthr05) [inline src/Math/src/ContactWrenchCone.cpp:115]:
```suggestion
    // adjoint transform brings the wrench expressed in the frame attached to the contact surface in
```

File: CHANGELOG.md (hunk 3 above)
Comment (GiulioRomualdi) [reply CHANGELOG.md:9]: Yes, I noticed but skew function is defined as "private" implementation in manif. I don't know if it is wise to directly their functions.


Let's ask the others @S-Dafarra and @traversaro

File: src/Math/include/BipedalLocomotion/Math/ContactWrenchCone.h (hunk 1 above)
Comment (GiulioRomualdi) [reply src/Math/include/BipedalLocomotion/Math/ContactWrenchCone.h]: Fixed :)

File: src/Math/include/BipedalLocomotion/Math/Skew.h (hunk 2 above)
Comment (GiulioRomualdi) [reply src/Math/include/BipedalLocomotion/Math/Skew.h]: Fixed here: https://github.com/dic-iit/bipedal-locomotion-framework/pull/352/commits/9a5c03370b1bdb9288711b7a0b8fca871571c3f8

**Comment by GiulioRomualdi** [comment]
> I would add some test cases where you test the constraint with values you are sure are inside, and values for which you are sure are outside. I have seen you have tested with respect to what you have computed in Matlab, but you could have done the same mistake twice. Also, if you need to edit the constraint by adding or removing lines, that part would break.

I agree on that @S-Dafarra, I'm working on this 😃

**Comment by GiulioRomualdi** [comment]
Here I improved the test https://github.com/dic-iit/bipedal-locomotion-framework/pull/352/commits/50e1f44afc2dc94166b51983914ab6ed61b3c00d

File: CHANGELOG.md (hunk 3 above)
Comment (traversaro) [reply CHANGELOG.md:9]: Are you sure that it is meant to be private? It is used in the manif examples (see https://github.com/artivis/manif/blob/5d1297b6d8996a5123aea6f1508cd8a029cac8e2/examples/se_2_3_localization.cpp#L368) and the `private` namespace is actually closed before `skew` is defined, see https://github.com/artivis/manif/blob/7c6b8b2dab6d940204136a5b280db1c243d72c4a/include/manif/impl/eigen.h#L129 .

Comment (GiulioRomualdi) [reply CHANGELOG.md:9]: Ok! Since the implementation that I wrote is actually  a copy-paste of the manif one I think I can remove the file

Comment (GiulioRomualdi) [reply CHANGELOG.md:9]: Removed!

**Comment by GiulioRomualdi** [comment]
@S-Dafarra and @prashanthr05 let me know if you're fine with the modification that I did``

File: src/Math/include/BipedalLocomotion/Math/ContactWrenchCone.h [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+{
+    Eigen::MatrixXd m_A;
+    Eigen::VectorXd m_b;
+
+    bool m_isIntialized{false}; /**< True if the class has been correctly initialize */
+
+public:
+
+    /**
+     * Initialize the continuous algebraic riccati equation solver.
```
Comment (S-Dafarra) [inline src/Math/include/BipedalLocomotion/Math/ContactWrenchCone.h:70]: Copy paste

**Review by S-Dafarra — APPROVED** [review]
Just a simple copy-paste error in the docs

**Review by prashanthr05 — APPROVED** [review]

**Pull request merged**